File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2018/Y18D15.kt
package org.gristle.pdxpuzzles.advent.y2018

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.objects.Coord
import org.gristle.pdxpuzzles.utilities.graph.Graph

class Y18D15(val input: String) : Day {

    /**
     * Abstract class for both elves and goblins. Responsible for tracking health, playing each players turn,
     * including attacking and moving
     */
    abstract class Player {
        var health = 200

        private val isDead: Boolean get() = health < 1

        /**
         * Finds the player at the specified other position and commands it to take damage.
         */
        private fun attack(world: World, otherPos: Coord) {
            world[otherPos]?.takeDamage(world, otherPos, damage(world))
        }

        /**
         * Looks for adjacent targets. If found, attack the one with lowest health. Otherwise, move to closest
         * target and attack if now adjacent.
         */
        fun playTurn(world: World, pos: Coord) {
            if (isDead) return

            val enemies = enemies(world)

            // utility function for finding adjacent targets to attack
            fun adjacentOpponentPos(pos: Coord) = pos
                .getNeighbors()
                .filter { it in enemies }
                .minByOrNull { enemies.getValue(it).health }

            // find adjacent target...
            val adjacentTarget = adjacentOpponentPos(pos)

            if (adjacentTarget != null) {
                attack(world, adjacentTarget) // ...and attack it
            } else { // if no adjacent target...
                val newPos = move(world, pos) // move one space to a target
                val adjacentTargetAfterMove = adjacentOpponentPos(newPos) // and attack if adjacent
                if (adjacentTargetAfterMove != null) attack(world, adjacentTargetAfterMove)
            }
        }

        /**
         * Takes damage and removes the player from the world if dead.
         */
        private fun takeDamage(world: World, pos: Coord, damage: Int) {
            health -= damage
            if (isDead) {
                friends(world).remove(pos)
            }
        }

        /**
         * Finds the closest spot adjacent to a target and move one step toward it. Uses BFS.
         */
        fun move(world: World, pos: Coord): Coord {
            val friends = friends(world)
            val enemies = enemies(world)

            // The coordinates that are adjacent to targets.
            val inRange: Set<Coord> = enemies
                .keys
                .flatMap { opponentPos -> opponentPos.getNeighbors().filter { world.canMove(it) } }
                .toSet()

            // Function to find neighboring coordinates for any given position
            val edges: (Coord) -> List<Coord> = { it.getNeighbors().filter { neighbor -> world.canMove(neighbor) } }

            // BFS takes a starting coordinate and spreads out to all legal moving positions, stopping when it
            // finds a target in range. It then reconstructs the path it took, finds the first step in that path,
            // and that is the new position.
            val newPos = Graph.bfsSequence(pos, defaultEdges = edges)
                .firstOrNull { (pos) -> pos in inRange }
                ?.path()
                ?.get(1)
                ?.id

            return if (newPos == null) {
                pos
            } else {
                // update the friends map and return the new position
                friends.remove(pos)
                friends[newPos] = this
                newPos
            }
        }

        abstract fun damage(world: World): Int

        abstract fun friends(world: World): MutableMap<Coord, Player>

        abstract fun enemies(world: World): MutableMap<Coord, Player>

    }

    class Elf : Player() {
        override fun friends(world: World) = world.elves
        override fun enemies(world: World) = world.goblins
        override fun damage(world: World) = world.elfDamage
    }

    class Goblin : Player() {
        override fun friends(world: World) = world.goblins
        override fun enemies(world: World) = world.elves
        override fun damage(world: World) = 3
    }

    // State object for the "World;" ie, where each Elf, Goblin, and wall is, as well as the dimensions of the map.
    // It also stores how much damage elves do in this world.
    data class World(
        val width: Int,
        val height: Int,
        private val walls: Set<Coord>,
        val elves: MutableMap<Coord, Player>,
        val goblins: MutableMap<Coord, Player>,
        val elfDamage: Int
    ) {
        private val initialElves = elves.size

        /**
         * Creates a copy of the world with rejuvenated elves and goblins, and with a new elfDamage amount.
         */
        fun clone(elfDamage: Int): World {
            val newElves: MutableMap<Coord, Player> = elves
                .entries
                .associateTo(mutableMapOf()) { (pos) -> pos to Elf() }
            val newGoblins: MutableMap<Coord, Player> = goblins
                .entries
                .associateTo(mutableMapOf()) { (pos) -> pos to Goblin() }

            return World(width, height, walls, newElves, newGoblins, elfDamage)
        }

        /**
         * Checks whether a coordinate would be a legal move by checking if it's within the dimensions, and whether
         * there is a wall or player in that position.
         */
        fun canMove(pos: Coord): Boolean = pos.x in 0 until width
                && pos.y in 0 until height
                && pos !in walls
                && pos !in elves
                && pos !in goblins

        fun elfHealth() = elves.values.sumOf { it.health }
        fun goblinHealth() = goblins.values.sumOf { it.health }

        /**
         * Returns all players in reading order.
         */
        fun players() = (elves.entries + goblins.entries).sortedBy { (pos) -> pos.asIndex(width) }

        /**
         * Returns true if elves have lost. Part 2 logic makes elves lose if even one elf dies.
         */
        fun elvesLose(): Boolean = elves.isEmpty() || (elfDamage > 3 && elves.size < initialElves)

        /**
         * Returns player at a position.
         */
        operator fun get(pos: Coord): Player? = elves[pos] ?: goblins[pos]

        companion object {

            /**
             * Parses initial world from input string.
             */
            fun from(input: String): World {
                val width = input.indexOf('\n')

                // The string with no newlines. 
                val flattenedInput = input.replace("\n", "")
                val height = flattenedInput.length / width
                val elves = mutableMapOf<Coord, Player>()
                val goblins = mutableMapOf<Coord, Player>()
                val walls = mutableSetOf<Coord>()
                for (y in 0 until height) {
                    for (x in 0 until width) {
                        val pos = Coord(x, y)
                        when (flattenedInput[y * width + x]) {
                            '#' -> walls.add(pos)
                            'E' -> elves[pos] = Elf()
                            'G' -> goblins[pos] = Goblin()
                        }
                    }
                }
                return World(width, height, walls, elves, goblins, 3)
            }
        }
    }

    // The initial world used by both parts.
    private val initialWorld = World.from(input)

    /**
     * State class holding the final Round info plus opponentHp, used to find the final score.
     */
    data class Game(val round: Round, val opponentHp: Int) {
        fun score(): Int = round.round * opponentHp
    }

    /**
     * Enum class says whether the game continues, or elves or goblins won.
     */
    enum class WinState { ELVES, GOBLINS, CONTINUE }

    /**
     * Round state class holding the round number and the current winstate.
     */
    data class Round(val round: Int = 0, val winState: WinState = WinState.CONTINUE)

    /**
     * Main game loop. Returns a game object that has enough info to determine whether the game needs to be run
     * again (in part 2) and to calculate the score.
     */
    fun solve(elfDamage: Int = 3): Game {

        // Make a working copy of the initial world.
        val world = initialWorld.clone(elfDamage)

        // Takes a Round state, and plays another Round
        fun playRound(round: Round): Round {

            // get all remaining players in reading order
            val players = world.players()

            // for each player...
            players.forEachIndexed { index, (pos, player) ->

                // ...play a turn
                player.playTurn(world, pos)

                // determine if there is a winner
                val winState = when {
                    world.elvesLose() -> WinState.GOBLINS
                    world.goblins.isEmpty() -> WinState.ELVES
                    else -> WinState.CONTINUE
                }

                // if there is a winner, note that in the return Round. Only advance the round if the last player
                // won the game
                if (winState != WinState.CONTINUE) {
                    val winRound = if (index == players.lastIndex) round.round + 1 else round.round
                    return Round(winRound, winState)
                }
            }

            // otherwise return a continuing round
            return round.copy(round = round.round + 1)
        }

        // sequence runs rounds in succession until a winner is found. When found, returns the game state.
        generateSequence(Round(), ::playRound)
            .first { (_, winState) -> winState != WinState.CONTINUE }
            .let { round ->
                val opponentHp = if (round.winState == WinState.ELVES) world.elfHealth() else world.goblinHealth()
                return Game(round, opponentHp)
            }
    }

    // Runs the game once and returns the score.
    override fun part1() = solve().score()

    // Sequence runs games with increasing elf damage until there is a game where the elves win. Returns score.
    override fun part2() = generateSequence(4) { elfDamage -> elfDamage + 1 }
        .map { elfDamage -> solve(elfDamage) }
        .first { game -> game.round.winState == WinState.ELVES }
        .score()
}

fun main() = Day.runDay(Y18D15::class)

//    Class creation: 5ms
//    Part 1: 224370 (140ms)
//    Part 2: 45539 (252ms)
//    Total time: 399ms

File: rust/advent/src/bin/2018/y18d15.rs
use advent::utilities::get_input::get_input;
use itertools::Itertools;
use rustc_hash::{FxHashMap, FxHashSet};
use utilities::{graphs::{bfs, EdgeInfo, PathInfo}, structs::{coord::Coord2U, stopwatch::{ReportDuration, Stopwatch}, str_grid::StrGrid}};
use utilities::enums::cardinals::Cardinal;

const READING_ORDER: [Cardinal; 4] = [
    Cardinal::North,
    Cardinal::West,
    Cardinal::East,
    Cardinal::South,
];

type Input = World;
type Output = usize;
type Pos = Coord2U;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum Race { Elf, Goblin }

#[derive(Clone, Copy, Debug)]
struct Player {
    race: Race,
    health: i16,
}

impl Player {
    fn new(race: Race) -> Self {
        Self { race, health: 200 }
    }
    
    fn is_dead(&self) -> bool {
        self.health <= 0
    }

    fn enemy_race(&self) -> Race {
        match self.race {
            Race::Elf => Race::Goblin,
            Race::Goblin => Race::Elf,
        }
    }

    fn attack(&self, world: &mut World, opponent_pos: Pos, killed: &mut FxHashSet<Pos>) {
        let opponent: &mut Player = world.players.get_mut(&opponent_pos).unwrap();
        let damage = match self.race {
            Race::Elf => world.elf_damage,
            Race::Goblin => 3,
        };
        opponent.health -= damage;
        if opponent.is_dead() {
            world.players.remove(&opponent_pos);
            killed.insert(opponent_pos);
        }
    }

    fn move_pos(&self, world: &mut World, pos: Pos) -> Option<Pos> {
        // The coordinates that are adjacent to targets.
        let opponent_race = match self.race {
            Race::Elf => Race::Goblin,
            Race::Goblin => Race::Elf,
        };
        let in_range: FxHashSet<Pos> = world.player_positions_by_race(opponent_race)
            .flat_map(|pos| {
                pos.adjacent(false).into_iter().filter(|&n| world.can_move(n))
            })
            .collect();
        let search: PathInfo<Pos, usize> = bfs(
            pos,
            |_: EdgeInfo<usize>, pos| {
                READING_ORDER.iter()
                    .map(|&dir| pos.move_direction(dir, 1).unwrap())
                    .filter(|&n| world.can_move(n))
                    .collect()
            },
            |_: EdgeInfo<_>, pos| in_range.contains(pos)
        );
        search.end_index.map(|end_index| search.path(end_index)[1].state)
    }

    fn play_turn(&self, world: &mut World, pos: Pos, killed: &mut FxHashSet<Pos>) -> Option<Pos> {
        if self.is_dead() { return None; }
        let enemies: Vec<Pos> = world.player_positions_by_race(self.enemy_race()).collect();

        if let Some(adjacent_target) = Self::adjacent_opponent_pos(pos, &enemies, world) {
            self.attack(world, adjacent_target, killed);
            None
        } else {
            let new_pos = self.move_pos(world, pos);
            if let Some(new_pos) = new_pos {
                if let Some(adjacent_target) = Self::adjacent_opponent_pos(new_pos, &enemies, world) {
                    self.attack(world, adjacent_target, killed);
                }
            }
            new_pos
        }
    }

    fn adjacent_opponent_pos(pos: Pos, enemies: &[Pos], world: &World) -> Option<Pos> {
        READING_ORDER.iter()
            .map(|&dir| pos.move_direction(dir, 1).unwrap())
            .filter(|it| enemies.contains(it))
            .min_by_key(|it| world.players.get(it).map(|p| p.health))
    }
}

#[derive(Clone, Debug)]
struct World {
    elf_damage: i16,
    initial_elves: usize,
    walls: FxHashSet<Pos>,
    players: FxHashMap<Pos, Player>,
    width: usize,
    height: usize,
}

impl World {
    fn player_positions_by_race(&self, race: Race) -> impl Iterator<Item = Pos> {
        self.players.iter()
            .filter(move |(_, player)| player.race == race)
            .map(|(&pos, _)| pos)
    }

    fn can_move(&self, pos: Pos) -> bool {
        (0..self.height).contains(&pos.x())
            && (0..self.width).contains(&pos.y())
            && !self.walls.contains(&pos)
            && !self.players.contains_key(&pos)
    }

    // Returns all players in reading order.
    fn positions_in_order(&self) -> impl Iterator<Item = Pos> {
        self.players.keys()
            .sorted_unstable_by_key(|&pos| pos)
            .copied()
    }

    fn elves_lose(&self) -> bool {
        let elves = self.player_positions_by_race(Race::Elf).count();
        elves == 0 || (self.elf_damage > 3 && elves < self.initial_elves)
    }
}

impl From<&str> for World {
    fn from(value: &str) -> Self {
        let grid = StrGrid::new(value).unwrap();
        let width = grid.width - 1;
        let height = grid.height;
        let mut walls = FxHashSet::default();
        let mut players = FxHashMap::default();
        for (index, &b) in grid.s.iter().enumerate() {
            let x = index % grid.width;
            let y = index / grid.width;
            let pos = Pos::new2d(x, y);
            match b {
                b'#' => { walls.insert(pos); },
                b'G' => { players.insert(pos, Player::new(Race::Goblin)); },
                b'E' => { players.insert(pos, Player::new(Race::Elf)); },
                _ => {},
            }
        }

        let initial_elves = players.values()
            .filter(|p| p.race == Race::Elf)
            .count();

        Self {
            elf_damage: 3,
            initial_elves,
            walls,
            players,
            width,
            height,
        }
    }
}

#[derive(Debug, Copy, Clone)]
struct Game {
    round: Round,
    remaining_health: i16,
}

impl Game {
    fn score(&self) -> usize {
        self.round.round_number * self.remaining_health as usize
    }
}

#[derive(Debug, Copy, Clone)]
struct Round {
    round_number: usize,
    win_state: WinState,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
enum WinState { Elves, Goblins, Continue }

fn play_game(mut world: World) -> Game {
    let mut round = Round { round_number: 0, win_state: WinState::Continue };
    while round.win_state == WinState::Continue {
        round = play_round(&mut world, round);
    }   
    let winner = match round.win_state {
        WinState::Elves => Race::Elf,
        WinState::Goblins => Race::Goblin,
        WinState::Continue => unreachable!(),
    };
    let remaining_health: i16 = world.players.values()
        .filter(|player| player.race == winner)
        .map(|player| player.health)
        .sum();

    Game { round, remaining_health }
}

#[allow(dead_code)]
fn rep(world: &World) -> String {
    let mut rep = String::new();
    for y in 0..world.height {
        for x in 0..world.width {
            let pos = Pos::new2d(x, y);
            if world.walls.contains(&pos) {
                rep.push('#');
            } else if !world.players.contains_key(&pos) {
                rep.push('.');
            } else {
                if let Some(player) = world.players.get(&pos) {
                    if player.race == Race::Elf {
                        rep.push('E');
                    } else {
                        rep.push('G');
                    }
                }
            }
        }
        rep.push('\n');
    }
    rep
}

fn play_round(world: &mut World, round: Round) -> Round {
    let mut killed = FxHashSet::default();
    for (index, pos) in world.positions_in_order().collect_vec()
        .into_iter()
        .enumerate()
    {
        if killed.contains(&pos) { continue; }
        let Some(player) = world.players.remove(&pos) else { continue; };
        let new_pos = player.play_turn(world, pos, &mut killed)
            .unwrap_or(pos);
        world.players.insert(new_pos, player);
        let win_state = if world.elves_lose() {
            WinState::Goblins
        } else if world.player_positions_by_race(Race::Goblin).count() == 0 {
            WinState::Elves
        } else {
            WinState::Continue
        };
        if win_state != WinState::Continue {
            let win_round = if index == world.players.len() - 1 {
                round.round_number + 1
            } else {
                round.round_number
            };
            return Round { round_number: win_round, win_state }
        }
    }
    Round { round_number: round.round_number + 1, win_state: WinState::Continue }
}

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(18, 15).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(input.clone()), stopwatch.lap().report());
    println!("2. {} ({})", part2(input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    input.into()
}

fn part1(world: Input) -> Output {
    play_game(world).score()
}

fn part2(initial_world: Input) -> Output {
    let mut min: i16 = 3;
    let mut max: i16 = 3;
    let mut score: Option<usize> = None;
    loop {
        let elf_damage = if score == None {
            max += 40;
            max
        } else {
            (max - min) / 2 + min
        };
        let mut world = initial_world.clone();
        world.elf_damage = elf_damage;
        let game = play_game(world);
        if game.round.win_state == WinState::Elves {
            if score.is_some() {
                max = elf_damage;
            }
            score = Some(game.score());
        } else {
            if score == None {
                min = 44;
                max += 40;
            } else {
                min = elf_damage + 1;
            }
        }
        if min + 1 == max {
            return score.unwrap();
        }
    }
}

#[test]
fn default() {
    let input = get_input(18, 15).unwrap();
    let input = parse_input(&input);
    assert_eq!(224370, part1(input.clone()));
    assert_eq!(45539, part2(input));
}

// Input parsed (56μs)
// 1. 224370 (27ms)
// 2. 45539 (16ms)
// Total: 44ms